### PR TITLE
Story/legend/550

### PR DESCRIPTION
**Pull request by Sybrand**
- Auto-layout of legend (pass list of legend items, and layout happens automatically)
- Optional configuration of pool and overflow size (was hitting limits on local machine when changing ui, due to slow queries)
- Several headers because I'm getting sick of ZAP scan errors.

### CODE REVIEW

**Review by andrea-williams — CHANGES_REQUESTED** [review]
Please fix bug where RDPS percentiles get cut off from legend box.

Also, Sonarcloud is reporting a code smell - not sure if it's a red herring

File: web/src/utils/d3.ts [hunk 1]
```diff
@@ -339,95 +339,134 @@ export const drawArea = <T>({
...
+
+      // Move icon and text to the correct location.
+      icon.attr('transform', translateIcon(legendData.shape, xOffset, yOffset, icon))
+      const iconWidth = icon.node()?.getBBox().width ?? 0
+      text.attr(
+        'transform',
+        'translate(' + (xOffset + iconTextPadding + iconWidth) + ', ' + yOffset + ')'
+      )
+    })
+  return ((data.length / numColumns) | 0) * lineHeight + lineHeight
```
Comment (Sybrand) [inline web/src/utils/d3.ts:469]: @andrea-williams fixed the display issue, was using number 4, instead of numColumns.

**Comment by Sybrand** [comment]
> Please fix bug where RDPS percentiles get cut off from legend box.
> 
> Also, Sonarcloud is reporting a code smell - not sure if it's a red herring

Fixed the display issue. @Kyubinhan can you give me an opinion on the sonarcloud smell - it's the only way I can get typescript to be happy - typescript is honestly a bit of a pain with svg.

File: web/src/utils/d3.ts [hunk 2]
```diff
@@ -339,95 +339,134 @@ export const drawArea = <T>({
...
+        .style('alignment-baseline', 'central')
+        .style('fill', legendData.color)
+        .style('font-size', '9px')
+        .text(legendData.text)
+
+      const icon = createIcon(legendItem, legendData)
+      // Calculte x offset using the remainder.
+      const xOffset = (i % numColumns) * columnWidth
+      // Calculate y offset using the quotient.
+      const yOffset = ((i / numColumns) | 0) * lineHeight
```
Comment (Kyubinhan) [inline web/src/utils/d3.ts:459]: I'd prefer `Math.tranc` for better readability

**Review by Kyubinhan — APPROVED** [review]

File: web/src/utils/d3.ts (hunk 2 above)
Comment (Sybrand) [reply web/src/utils/d3.ts:459]: Sure thing - f.y.i. (not that it matters in this case - but | 0 is faster)

**Comment by Kyubinhan** [comment]
> > Please fix bug where RDPS percentiles get cut off from legend box.
> > Also, Sonarcloud is reporting a code smell - not sure if it's a red herring
> 
> Fixed the display issue. @Kyubinhan can you give me an opinion on the sonarcloud smell - it's the only way I can get typescript to be happy - typescript is honestly a bit of a pain with svg.

sorry. just saw this. Let me see if I can resolve this

File: web/src/features/fireWeather/components/graphs/PrecipGraph.tsx [hunk 3]
```diff
@@ -308,6 +283,42 @@ const PrecipGraph: React.FunctionComponent<Props> = ({
...
+    const svgElement = svgRef.current
+    if (svgElement) {
+      const svg = d3.select(svgElement)
+      // Grab the legend.
+      const legend = svg.select('.legend') as d3.Selection<
+        SVGGElement,
+        unknown,
+        null,
+        undefined
+      >
```
Comment (Kyubinhan) [inline web/src/features/fireWeather/components/graphs/PrecipGraph.tsx:314]:
```suggestion
      const legend = svg.select<SVGGElement>('.legend')
```

Comment (Sybrand) [reply web/src/features/fireWeather/components/graphs/PrecipGraph.tsx]: Awesome - thanks.

**Review by andrea-williams — APPROVED** [review]

**Pull request merged**